website_sale_coupon_page/controllers/main.py

```diff
@@ -0,0 +1,37 @@
+# Copyright 2021 Tecnativa - Carlos Roca
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import http
+from odoo.http import request
+
+
+class WebsiteSale(http.Controller):
+    @http.route(["""/promotions"""], type="http", auth="public", website=True)
+    def promotion(self, **post):
+        active_user_id = (
+            request.env["res.users"]
+            .sudo()
+            .search([("id", "=", request.env.context.get("uid"))])
+        )
+        active_partner_id = (
+            request.env["res.partner"]
+            .sudo()
+            .search([("id", "=", active_user_id.partner_id.id)])
+        )
+        all_promos = (
+            request.env["sale.coupon.program"]
+            .sudo()
+            .search(
+                [
+                    ("is_published", "=", True),
+                    "|",
+                    ("website_id", "=", False),
+                    ("website_id", "=", request.env.context.get("website_id")),
+                ]
+            )
+        )
+        promos = request.env["sale.coupon.program"]
+        for promo in all_promos:
+            if promo._is_valid_partner(active_partner_id):
```

Comment:
```suggestion
        all_promos = (
            request.env["sale.coupon.program"]
            .with_user(request.env.user)
            .search(
                [
                    ("is_published", "=", True),
                    "|",
                    ("website_id", "=", False),
                    ("website_id", "=", request.env["website"].get_current_website()),
                ]
            )
        )
        promos = request.env["sale.coupon.program"]
        for promo in all_promos:
            if promo._is_valid_partner(request.env.user.partner_id):
```
I don't really sure about all suggestion but probably simplify and improve code if some user logged show only allowed (if some addon filter promos).